Mathlib/Data/Nat/Factorial/NatCast.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir, María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández
+-/
+
+import Mathlib.Algebra.Algebra.Defs
+import Mathlib.Algebra.CharP.Defs
+import Mathlib.Algebra.Order.Field.Rat
+import Mathlib.Data.Int.GCD
+import Mathlib.Data.Nat.Factorial.Basic
+import Mathlib.Data.Nat.Prime.Basic
+
+/-!
+# Invertibility of factorials
+
+This file contains lemmas providing sufficient conditions for the cast of `n!` to a commutative
+(semi)ring `A` to be a unit.
+
+-/
+
+namespace IsUnit
+
+open Nat
+section Semiring
+
+variable {A : Type*} [Semiring A]
+
+theorem natCast_factorial_of_le {n : ℕ} (hn_fac : IsUnit (n ! : A))
+    {m : ℕ} (hmn : m ≤ n) : IsUnit (m ! : A) := by
+  obtain ⟨k, rfl⟩ := exists_add_of_le hmn
+  clear hmn
+  induction k generalizing m with
+  | zero => simpa using hn_fac
+  | succ k ih =>
+    rw [← add_assoc, add_right_comm] at hn_fac
+    have := ih hn_fac
+    rw [Nat.factorial_succ, Nat.cast_mul, Nat.cast_commute _ _ |>.isUnit_mul_iff] at this
+    exact this.2
+
+theorem natCast_factorial_of_lt {n : ℕ} (hn_fac : IsUnit ((n - 1)! : A))
+    {m : ℕ} (hmn : m < n) : IsUnit (m ! : A) :=
+  hn_fac.natCast_factorial_of_le <| le_sub_one_of_lt hmn
+
+theorem natCast_factorial_of_ratAlgebra [Algebra ℚ A] (n : ℕ) : IsUnit (n ! : A) := by
+  rw [← map_natCast (algebraMap ℚ A)]
+  apply IsUnit.map
+  simp [isUnit_iff_ne_zero, n.factorial_ne_zero]
```

Comment:
Here's a version that works for both `Rat` and `NNRat`:
```lean
/-- If `A` is an algebra over a characteristic-zero (semi)field, then `n!` is a unit. -/
theorem natCast_factorial_of_algebra (K : Type*) [Semifield K] [CharZero K] [Algebra K A] (n : ℕ) : IsUnit (n ! : A) := by
  rw [← map_natCast (algebraMap K A)]
  apply IsUnit.map
  simp [isUnit_iff_ne_zero, n.factorial_ne_zero]
```

Reply:
But this requires `Semifield` (which for instance is not true for `Polynomial Rat`).

Reply:
Sure it is, set `K := Rat`. My proposal is to replace callers of `natCast_factorial_of_ratAlgebra` with `natCast_factorial_of_algebra Rat`.

Reply:
Ah, of course you are right, sorry (I had not read carefully).